doc/SAI_generate_debug_dump/SAI_generate_debug_dump.md

```diff
@@ -0,0 +1,262 @@
+
+Currently, the SAI dump file is generated only during SAI failures by executing a dedicated executable named "saisdkdump" (which linkage with the SAI lib and during initialization it creates a new switch in redundant mode) 
+
+This new feature allows users to generate a SAI debug dump file using `show tech-support` command, not necessarily during failure.
+
+### Requirements
+
++ Add infrastructure to generate a SAI debug dump file upon user request
++ generate a SAI debug dump file from 'show techsupport' command.
++ Generate a SAI debug dump file within the context of Syncd.
++ Maintain the existing mechanism for generating the SAI debug dump file on failure.
+
+### Assumptions
+
+The SAI global API `sai_dbg_generate_dump` operates as a blocking function.
+
+### Architecture Design
+
+Adding a new infrastructure without changes in existing Sonic Architecture
+
+1. A user command, such as `show techsupport` triggers the `generate_sai_dump`, and creates a new table with the dump file name to create in the APPL DB.
+2. A new orchestration agent, `DbgGenDumpOrch`, is triggered to handle the request.
+3. `DbgGenDumpOrch` writes the file name to the ASIC DB and sets a new operation `REDIS_ASIC_STATE_COMMAND_DBG_GEN_DUMP` for syncd.
+4. Syncd calls the global SAI API `dbgGenerateDump` to generate the debug dump file, which is saved in syncd's file system.
+5. Syncd sends a reply back to `DbgGenDumpOrch`.
+6. `DbgGenDumpOrch` analyzes the response.
+7. `DbgGenDumpOrch` updates the result in the APPL DB.
+8. The user command retrieves the result.
+9. The debug dump file is pulled on success.
+
+The below diagram explains the generate debug dump file flow
+
+![Architecture Design](images/generate_debug_dump_file.png)
+
+
+### Implementation
+
+#### generate_sai_dump bash script
+Introduced a new script `/usr/local/bin/gen_sai_dbg_dump.sh`
+
+```
+###############################################################################
+# generate_sai_dump
+#
+# Description:
+#  This function
+#  it ensures that the `syncd` container is running before initiating the dump.
+#  triggers the generation of a SAI debug dump file through Redis APPL DB.
+#  it waits for the file by Polling (with timeout) the APPL DB for the result.
+#  it removes the table from the DB when done.
+#
+# Arguments:
+#  $1 - Filename for the SAI debug dump file.
+#  $2 - Optional timeout for file readiness (default: 60 seconds).
+#
+# Returns:
+#  0 - On success
+#  1 - On failure
+###############################################################################
+generate_sai_dump() {
+
+}
+```
+
+ The script also can be invoked from the CLI to generate the dump file directly under the given name (without calling `show techsupport` command):
+```
+/usr/local/bin/gen_sai_dbg_dump.sh -f /tmp/my_dump_file.log
+```
+
+#### show techsupport
+Introduced a new generic API, `generate_sai_dbg_dump_file`, in `generate_dump.sh` (invoked by the `show techsupport` command) to create a debug dump file:
```

Comment:
For nvidia platform, sai_dbg_generate_dump call generates a bunch of files. 

Eg: 
```
docker exec syncd mkdir -p /tmp/saisdkdump/
docker exec syncd saisdkdump -f /tmp/saisdkdump/sai_sdk_dump_12_10_2024_09_21_PM

ls -l /tmp/saisdkdump/
total 60948
-rw-r--r-- 1 root root 11854441 Dec 10 21:25 sai_sdk_dump_12_10_2024_09_21_PM
-rw-r--r-- 1 root root     1826 Dec 10 21:25 sai_sdk_dump_12_10_2024_09_21_PM.json
drwxr-xr-x 2 root root     4096 Dec 10 21:25 sai_sdk_dump_12_10_2024_09_21_PM_extras
-rw-r--r-- 1 root root     1137 Dec 10 21:25 sdk_dump_ext_.....summary.txt
-rw-r--r-- 1 root root  1201437 Dec 10 21:25 sdk_dump_ext_.....driver.txt
...............
```

I'm not aware of other vendors but i believe SONiC should have flexibility to provide a path for vendor SAI to dump whatever it deems relevant.  Thus, i recommend the following, let me know what you think

- Update generate_sai_dbg_dump_file to take in two parameters
Eg: `generate_sai_dbg_dump_file -p "$path" -f "$filename"`
- Make sure syncd runs sai_dbg_generate_dump with arg $path/$filename
- After the execution is complete, techsupport process will copy the contents under $path back into the host and save them into the tar file.

Reply:
I initially assumed that the SAI API generates a single file since it receives a specific file name rather than just a path. However, I have now realized that the API may generate multiple files. 
To accommodate this, I will modify the implementation so that the`show tech-support` process copies all files from the specified directory instead of targeting a single file.  
In my opinion, Since the `generate_sai_dbg_dump_file` API is generic for all vendors and is invoked as part of the main `show tech-support` flow, the path for the SAI to place the files in the syncd container remains generic as well, using `/var/log/dbg_gen_dump.log`. The `generate_sai_dump` function in the `gen_sai_dbg_dump.sh` script ensures that the `/var/log/` directory exists and creates it if necessary. This directory is temporary, as the files generated are moved from the `syncd` filesystem to the host machine once the process completes. Therefore, I don’t see any value in defining separate paths for different vendors.  
Ultimately, since the purpose of calling `generate_sai_dbg_dump_file` is to generate the dump files and copy them to the show tech-support folder, the temporary paths where the files are stored in between do not really matter.
To avoid potential conflicts, I will create a unique directory under `/var/log/` to store the files temporarily, ensuring that such a directory does not already exist.

Reply:
Having a standard directory for everyone should suffice and the thread that collects the debug dump should be made singleton. i.e. only one instance runs at a time. 

I think after compressing and moving the contents, we should clear the folder.